src/core_plugins/kibana/public/visualize/wizard/type_selection/type_selection.tsx

```diff
@@ -0,0 +1,215 @@
+                  <NewVisHelp />
+                </React.Fragment>
+              )}
+            </EuiFlexItem>
+          </EuiFlexGroup>
+        </div>
+      </React.Fragment>
+    );
+  }
+
+  private filteredVisTypes(visTypes: VisType[], query: string): VisTypeListEntry[] {
+    const types = visTypes.filter(type => {
+      // Filter out all lab visualizations if lab mode is not enabled
+      if (!this.props.showExperimental && type.stage === 'experimental') {
+        return false;
+      }
+
+      // Filter out hidden visualizations
+      if (type.hidden) {
+        return false;
+      }
+
+      return true;
+    });
+
+    let entries: VisTypeListEntry[];
+    if (!query) {
+      entries = types.map(type => ({ ...type, highlighted: false }));
+    } else {
+      const q = query.toLowerCase();
+      entries = types.map(type => {
+        const matchesQuery =
+          type.name.toLowerCase().includes(q) ||
+          type.title.toLowerCase().includes(q) ||
+          (typeof type.description === 'string' && type.description.toLowerCase().includes(q));
+        return { ...type, highlighted: matchesQuery };
+      });
+    }
+
+    return sortByOrder(entries, ['highlighted', 'title'], ['desc', 'asc']);
+  }
+
+  private renderVisType = (visType: VisTypeListEntry) => {
+    let stage = {};
+    if (visType.stage === 'experimental') {
+      stage = {
+        betaBadgeLabel: i18n.translate('kbn.visualize.newVisWizard.experimentalTitle', {
+          defaultMessage: 'Experimental',
+        }),
+        betaBadgeTooltipContent: i18n.translate('kbn.visualize.newVisWizard.experimentalTooltip', {
+          defaultMessage: 'This visualization is experimental.',
+        }),
+      };
+    }
+    const isDisabled = this.state.query !== '' && !visType.highlighted;
+    const legacyIconClass = classnames(
+      'kuiIcon',
+      'visNewVisDialog__typeLegacyIcon',
+      visType.legacyIcon
+    );
+    return (
+      <EuiKeyPadMenuItemButton
+        key={visType.name}
+        label={<span data-test-subj="visTypeTitle">{visType.title}</span>}
+        onClick={() => this.props.onVisTypeSelected(visType)}
+        onFocus={() => this.highlightType(visType)}
+        onMouseEnter={() => this.highlightType(visType)}
+        onMouseLeave={() => this.highlightType(null)}
+        onBlur={() => this.highlightType(null)}
+        className="visNewVisDialog__type"
+        data-test-subj={`visType-${visType.name}`}
+        data-vis-stage={visType.stage}
+        disabled={isDisabled}
+        {...stage}
+      >
+        {visType.image && (
+          <img src={visType.image} aria-hidden="true" className="visNewVisDialog__typeImage" />
+        )}
+        {!visType.image && visType.legacyIcon && <span className={legacyIconClass} />}
+        {!visType.image &&
+          !visType.legacyIcon && (
+            <EuiIcon type={visType.icon} size="l" color="secondary" aria-hidden="true" />
+          )}
```

Comment:
had to read through the conditionals a couple times here, I wonder if there's any way to clean this up?

Reply:
I extracted it into it's own component with a better documentation. Also one of these path will be removed in 7.0, making this way more readable already :)